server/src/main.rs

```diff
@@ -149,31 +150,67 @@ async fn main() -> Result<(), anyhow::Error> {
+                    }
+                    _ = validator.schedule_revalidation_of_startup_tokens(edge.tokens, lazy_feature_refresher) => {
+                        tracing::info!("Token validator validation of startup tokens was unexpectedly shut down");
+                    }
+                    _ = metrics_pusher::prometheus_remote_write(prom_registry_for_write, edge.prometheus_remote_write_url, edge.prometheus_push_interval, edge.prometheus_username, edge.prometheus_password, app_name) => {
+                        tracing::info!("Prometheus push unexpectedly shut down");
+                    }
                 }
-                _ = metrics_pusher::prometheus_remote_write(prom_registry_for_write, edge.prometheus_remote_write_url, edge.prometheus_push_interval, edge.prometheus_username, edge.prometheus_password, app_name) => {
-                    tracing::info!("Prometheus push unexpectedly shut down");
+            } else {
+                tokio::select! {
+                    _ = server.run() => {
+                        tracing::info!("Actix is shutting down. Persisting data");
+                        clean_shutdown(persistence.clone(), lazy_feature_cache.clone(), lazy_token_cache.clone(), metrics_cache_clone.clone(), feature_refresher.clone()).await;
+                        tracing::info!("Actix was shutdown properly");
+                    },
+                    _ = refresher.start_refresh_features_background_task() => {
+                        tracing::info!("Feature refresher unexpectedly shut down");
+                    }
+                    _ = unleash_edge::http::background_send_metrics::send_metrics_task(metrics_cache_clone.clone(), refresher.clone(), edge.metrics_interval_seconds.try_into().unwrap()) => {
+                        tracing::info!("Metrics poster unexpectedly shut down");
+                    }
+                    _ = persist_data(persistence.clone(), lazy_token_cache.clone(), lazy_feature_cache.clone()) => {
+                        tracing::info!("Persister was unexpectedly shut down");
+                    }
+                    _ = validator.schedule_validation_of_known_tokens(edge.token_revalidation_interval_seconds) => {
+                        tracing::info!("Token validator validation of known tokens was unexpectedly shut down");
+                    }
+                    _ = validator.schedule_revalidation_of_startup_tokens(edge.tokens, lazy_feature_refresher) => {
+                        tracing::info!("Token validator validation of startup tokens was unexpectedly shut down");
+                    }
+                    _ = metrics_pusher::prometheus_remote_write(prom_registry_for_write, edge.prometheus_remote_write_url, edge.prometheus_push_interval, edge.prometheus_username, edge.prometheus_password, app_name) => {
+                        tracing::info!("Prometheus push unexpectedly shut down");
+                    }
```

Comment:
Wish we could simplify this and reduce the duplicated code but I don't have any better ideas atm.

Reply:
Yeah, I know. I hate this. But we can refactor it later. I don't think this'll be the final version (I hope not!).